[AMDGPU] Support wide register or subregister access when emitting s_singleuse_vdst instructions.

ScottEgerton opened the pull request:
Both single use producer and consumer instructions using wide/sub
registers are now correctly tracked and eligible for being marked as
single use.

ScottEgerton committed:
[AMDGPU] Support wide register or subregister access when emitting s_singleuse_vdst instructions.

Both single use producer and consumer instructions using wide/sub
registers are now correctly tracked and eligible for being marked as
single use.

jayfoad commented on llvm/test/CodeGen/AMDGPU/insert-singleuse-vdst.mir:
```diff
@@ -398,7 +398,6 @@ body: |
   ; CHECK-NEXT:   successors: %bb.1(0x80000000)
   ; CHECK-NEXT:   liveins: $sgpr0_sgpr1
   ; CHECK-NEXT: {{  $}}
-  ; CHECK-NEXT:   S_SINGLEUSE_VDST 1
```
Why did this change? I don't see any wide register or subregister accesses related to the def of vgpr0 in this test.

ScottEgerton replied:
I should have mentioned this in the commit message. This test should not be marked as single use as the use is after an exec mask change. This is caused by the change in AMDGPUInsertSingleUseVDST.cpp on line 110 replacing the RegisterUseCount.erase(Reg) with RegisterUseCount[Unit] = 0. When an exec mask change is detected, all registers in RegisterUseCount are given a value of 2 uses to exclude them from being marked as single use producers (line 92). However, when the values were being erased, they were no longer being correctly excluded. Setting the use count to 0 and keeping them in RegisterUseCount prevents this.

jayfoad replied:
Could you split that fix out into a separate PR please?

ScottEgerton replied:
Sure, no problem.

ScottEgerton committed:
fixup! [AMDGPU] Support wide register or subregister access when emitting s_singleuse_vdst instructions.

Change-Id: If41aa496d852d0ac1fddfc75f19acbfa20d2309e

ScottEgerton committed:
Separate out exec mask fix

Changing erasing unused registers to setting them
to 0 fixes an issue where single use values were
being incorrectly detected across exec mask
changes. However this is not relevant to wise/sub
register changes and will be included
in a different patch.

jayfoad commented on llvm/lib/Target/AMDGPU/AMDGPUInsertSingleUseVDST.cpp:
```diff
@@ -64,13 +64,15 @@ class AMDGPUInsertSingleUseVDST : public MachineFunctionPass {
-      for (const auto &Liveouts : MBB.liveouts())
-        RegisterUseCount[Liveouts.PhysReg] = 2;
+      for (const auto &Liveout : MBB.liveouts()) {
+        for (const MCRegUnit &Unit : TRI->regunits(Liveout.PhysReg))
```
Really this ought to use `MCRegUnitMaskIterator` to iterate over only the regunits of `Liveout.PhysReg` that are enabled in `Liveout.LaneMask`. See other uses of `MCRegUnitMaskIterator` in `lib/CodeGen/` for examples. Or you can just leave a TODO if you don't want to implement that now.

ScottEgerton replied:
I wasn't aware of `MCRegUnitMaskIterator`. I've added this in.

ScottEgerton committed:
Make use of MCRegUnitMaskIterator

This allows checking that liveout registers are
enabled in Liveout.LaneMask before excluding them from single use
optimisations.

jayfoad reviewed (approved):
LGTM.
